08-read-write-plot.qmd

```diff
@@ -291,6 +292,44 @@ The next two sections will demonstrate how to do this.
 
 ### Vector data
 
+The counterpart of `gpd.read_file` is the `.to_file` method that a `GeoDataFrame` has. It allows you to write `GeoDataFrame` objects to a wide range of geographic vector file formats, including the most common, such as `.geojson`, `.shp` and `.gpkg`. Based on the file name, `.to_file` decides automatically which driver to use. The speed of the writing process depends also on the driver.
+
+```{python}
+world.to_file('output/world.gpkg')
+```
+
+Note: if you try to write to the same data source again, the function will overwrite the file:
+
+```{python}
+world.to_file('output/world.gpkg')
+```
+
+Instead of overwriting the file, we could add a new layer to the file with `mode='a'` ("append" mode, as opposed to the default `mode='w'` for "write" mode). Appending is supported by several spatial formats, including GeoPackage. For example:
+
+```{python}
+world.to_file('output/world_many_features.gpkg')
+world.to_file('output/world_many_features.gpkg', mode='a')
+```
+
+Here, `world_many_features.gpkg` will contain a polygonal layer named `world` with two "copies" of each country (that is 177×2=354 features, whereas the `world` layer has 177 features).
+
+Alternatively, you can create another, separate, layer, within the same file. The GeoPackage format also supports multiple layers within one file. For example:
+
+```{python}
+world.to_file('output/world_many_layers.gpkg')
+world.to_file('output/world_many_layers.gpkg', layer='world2')
```

Comment:
This is really interesting and goes beyond the R version as far as I recall. Interestingly the layers don't have to have anything in common as far as I understand. Could you even have raster and vector data in the same .gpkg file with different layer names?

Reply:
Yes, but the support for raster data is very limited in .gpkg (not recommend).